lib/Common/DataStructures/QuickSort.h

```diff
@@ -6,45 +6,277 @@
+    };
+
+    union CC_QSORT_SWAP2
+    {
+        char chr[2]; // others
+    };
+
+    #define CC_QSORT_SWAP_LOOP(T, a, b, nsize)        \
+    {                                       \
+        for (size_t i = 0; i < nsize; i++)  \
+        {                                   \
+            T c = (a)[i];                   \
+            (a)[i] = (b)[i];                \
+            (b)[i] = c;                     \
+        }                                   \
+    }
+
+    template <class Policy, class T>
+    class QuickSortSwap
     {
     public:
-        static void Sort(T* low, T* high, Comparer comparer, void* context)
+        static inline void swap(T* a, T* b, size_t size)
         {
-            if ((low == NULL)||(high == NULL))
+            if (a == b) return;
+
+            size_t tsize;
+            if (size >= 8)
             {
-                return;
+                tsize = 8;
+                size_t nsize = size / tsize;
+                CC_QSORT_SWAP_LOOP(CC_QSORT_SWAP8, (CC_QSORT_SWAP8*) a, (CC_QSORT_SWAP8*) b, nsize)
+            }
+            else if (size >= 4)
+            {
+                tsize = 4;
+                size_t nsize = size / tsize;
+                CC_QSORT_SWAP_LOOP(CC_QSORT_SWAP4, (CC_QSORT_SWAP4*) a, (CC_QSORT_SWAP4*) b, nsize)
             }
-            if (high > low)
+            else
             {
-                T* pivot = Partition(low, high, comparer, context);
-                Sort(low, pivot-1, comparer, context);
-                Sort(pivot+1, high, comparer, context);
+                tsize = 2;
+                size_t nsize = size / tsize;
+                CC_QSORT_SWAP_LOOP(CC_QSORT_SWAP2, (CC_QSORT_SWAP2*) a, (CC_QSORT_SWAP2*) b, nsize)
             }
+
+            CC_QSORT_SWAP_LOOP(char, (char*) a, (char*) b, size % tsize);
         }
+    };
 
-    private:
-        static T* Partition(T* l, T* r, Comparer comparer, void* context)
+#ifdef RECYCLER_WRITE_BARRIER
+    template <class T>
+    class QuickSortSwap<_write_barrier_policy, T>
+    {
+    public:
+        static inline void swap(T* a, T* b, size_t _)
         {
-            T* i = l-1;
-            T* j = r;
-            for (;;)
-            {
-                while (comparer(context, ++i, r) < 0) ;
-                while (comparer(context, r, --j) < 0) if (j == l) break;
-                if (i >= j) break;
-                swap(*i, *j);
-            }
-            swap(*i, *r);
-            return i;
+            AssertMsg(_ == 1, "No byte to byte copies when a type has SWB is enabled. size should be 1!");
+            T temp = *a;
+            *a = *b;
+            *b = temp;
         }
+    };
+#endif
 
-        inline static void swap(T& x, T& y)
+    template <class Policy, class T, class Comparer>
+    class QuickSort
+    {
+    public:
+        #define CCQ_SWAP QuickSortSwap<Policy, T>::swap
+        static void Sort(T* base, size_t nmemb, size_t size, Comparer comparer, void* context)
```

Comment:
`write_barrier_policy` swap is a basic `=` operator. Also, the algorithm doesn't cast / iterate over swb enabled array. That's why  `sizeof(T) wouldn't apply to write_barrier_policy`. However I could iterate `T` type instead of `char` for other non_write_barrier_policy types. In this case, code would have to cast back and forward on swap.